### PR TITLE
Reduce the number of objects allocations

### PR DESCRIPTION
**Pull request by kajetanowicz**
The existing implementation of this gem allocates excessive number of objects. In order to reduce the memory footprint, I profiled the code by using the following snippet: 

```ruby
require 'structural'
require 'ruby-prof'

class StructuralModel
  include Structural::Model

  field :field1
  field :field2
  field :field3
  field :field4
  field :field5
  field :field6
  field :field7
  field :field8
  field :field9
  field :field10
end

simple_hash = (1..10).each_with_object(Hash.new) do |number, hash|
  hash["field#{number}"] = number.to_s
end

RubyProf.measure_mode = RubyProf::ALLOCATIONS
iterations = 10000

report = RubyProf.profile do

  iterations.times do
    model = StructuralModel.new(simple_hash)

    model.field1
    model.field2
    model.field3
    model.field4
    model.field5
    model.field6
    model.field7
    model.field8
    model.field9
    model.field10
  end

end

report.eliminate_methods!([/Integer#times/, /field/])
printer = RubyProf::GraphPrinter.new(report)
printer.print($stdout)
```

The result of this experiment revealed that by accessing 10000 models each containing 10 fields current implementation allocates about `1160000` objects:

```
Measure Mode: allocations
Thread ID: 4643180
Fiber ID: 12571800
Total Time: 1160011.0
Sort by: total_time

  %total   %self      total       self       wait      child            calls     name
--------------------------------------------------------------------------------
 100.00%   0.00%1160011.000      1.000      0.0001160010.000                1     Global#[No method]
                 910010.000      0.000      0.000 910010.000    100000/200000     Proc#call
                 250000.000  10000.000      0.000 240000.000     10000/120000     Class#new
--------------------------------------------------------------------------------
                 910010.000 110000.000      0.000 800010.000    100000/100000     Proc#call
  78.45%   9.48% 910010.000 110000.000      0.000 800010.000           100000     BasicObject#instance_eval
                 800010.000 200000.000      0.000 600010.000    100000/100000     Structural::Model#memoize
--------------------------------------------------------------------------------
                 400000.000      0.000      0.000 400000.000    100000/200000     Structural::Model#memoize
                 910010.000      0.000      0.000 910010.000    100000/200000     Global#[No method]
  78.45%   0.00% 910010.000      0.000      0.000 910010.000           200000    *Proc#call
                 910010.000 110000.000      0.000 800010.000    100000/100000     BasicObject#instance_eval
                 400000.000      0.000      0.000 400000.000    100000/100000     Structural::Model::Field#value_of
--------------------------------------------------------------------------------
                 800010.000 200000.000      0.000 600010.000    100000/100000     BasicObject#instance_eval
  68.97%  17.24% 800010.000 200000.000      0.000 600010.000           100000     Structural::Model#memoize
                 400000.000      0.000      0.000 400000.000    100000/200000     Proc#call
                 200000.000 200000.000      0.000      0.000    200000/300000     Symbol#to_s
                     10.000     10.000      0.000      0.000    100000/100000     Kernel#instance_variable_set
                      0.000      0.000      0.000      0.000    100000/100000     Kernel#instance_variable_get
--------------------------------------------------------------------------------
                 400000.000      0.000      0.000 400000.000    100000/100000     Proc#call
  34.48%   0.00% 400000.000      0.000      0.000 400000.000           100000     Structural::Model::Field#value_of
                 200000.000      0.000      0.000 200000.000    100000/100000     Structural::Model::Field#default_value
                 100000.000      0.000      0.000 100000.000    100000/100000     Structural::Model::Field#cast
                 100000.000      0.000      0.000 100000.000    100000/100000     Structural::Model::Field#key
                      0.000      0.000      0.000      0.000    100000/400000     Hash#fetch
--------------------------------------------------------------------------------
                  10000.000  10000.000      0.000      0.000     10000/120000     <Class::Structural::Hashifier>#hashify
                 100000.000 100000.000      0.000      0.000    100000/120000     <Module::Structural::Model::TypeCasts>#cast
                 250000.000  10000.000      0.000 240000.000     10000/120000     Global#[No method]
  30.17%   9.48% 350000.000 110000.000      0.000 240000.000           120000    *Class#new
                 240000.000      0.000      0.000 240000.000      10000/10000     Structural::Model#initialize
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model::TypeCasts::Cast#initialize
                      0.000      0.000      0.000      0.000      10000/10000     Structural::Hashifier#initialize
--------------------------------------------------------------------------------
                 100000.000 100000.000      0.000      0.000    100000/300000     Structural::Model::Field#key
                 200000.000 200000.000      0.000      0.000    200000/300000     Structural::Model#memoize
  25.86%  25.86% 300000.000 300000.000      0.000      0.000           300000     Symbol#to_s
--------------------------------------------------------------------------------
                 240000.000      0.000      0.000 240000.000      10000/10000     Structural::Model#initialize
  20.69%   0.00% 240000.000      0.000      0.000 240000.000            10000     <Class::Structural::Hashifier>#hashify
                 230000.000      0.000      0.000 230000.000      10000/10000     Structural::Hashifier#hashify
                  10000.000  10000.000      0.000      0.000     10000/120000     Class#new
--------------------------------------------------------------------------------
                 240000.000      0.000      0.000 240000.000      10000/10000     Class#new
  20.69%   0.00% 240000.000      0.000      0.000 240000.000            10000     Structural::Model#initialize
                 240000.000      0.000      0.000 240000.000      10000/10000     <Class::Structural::Hashifier>#hashify
--------------------------------------------------------------------------------
                 230000.000      0.000      0.000 230000.000      10000/10000     <Class::Structural::Hashifier>#hashify
  19.83%   0.00% 230000.000      0.000      0.000 230000.000            10000     Structural::Hashifier#hashify
                 220000.000  20000.000      0.000 200000.000      10000/10000     Enumerable#map
                  10000.000  10000.000      0.000      0.000      10000/10000     <Class::Hash>#[]
--------------------------------------------------------------------------------
                 220000.000  20000.000      0.000 200000.000      10000/10000     Structural::Hashifier#hashify
  18.97%   1.72% 220000.000  20000.000      0.000 200000.000            10000     Enumerable#map
                 200000.000 200000.000      0.000      0.000      10000/10000     Hash#each
--------------------------------------------------------------------------------
                 200000.000 200000.000      0.000      0.000    100000/100000     Structural::Model::Field#default_value
  17.24%  17.24% 200000.000 200000.000      0.000      0.000           100000     Kernel#proc
--------------------------------------------------------------------------------
                 200000.000      0.000      0.000 200000.000    100000/100000     Structural::Model::Field#value_of
  17.24%   0.00% 200000.000      0.000      0.000 200000.000           100000     Structural::Model::Field#default_value
                 200000.000 200000.000      0.000      0.000    100000/100000     Kernel#proc
--------------------------------------------------------------------------------
                 200000.000 200000.000      0.000      0.000      10000/10000     Enumerable#map
  17.24%  17.24% 200000.000 200000.000      0.000      0.000            10000     Hash#each
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Hashifier#as_data
                      0.000      0.000      0.000      0.000    100000/100000     String#to_s
--------------------------------------------------------------------------------
                 100000.000      0.000      0.000 100000.000    100000/100000     Structural::Model::Field#cast
   8.62%   0.00% 100000.000      0.000      0.000 100000.000           100000     <Module::Structural::Model::TypeCasts>#cast
                 100000.000 100000.000      0.000      0.000    100000/120000     Class#new
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model::TypeCasts::Identity#cast
                      0.000      0.000      0.000      0.000    100000/400000     Hash#fetch
                      0.000      0.000      0.000      0.000    100000/100000     <Module::Structural::Model::TypeCasts>#casts
--------------------------------------------------------------------------------
                 100000.000      0.000      0.000 100000.000    100000/100000     Structural::Model::Field#value_of
   8.62%   0.00% 100000.000      0.000      0.000 100000.000           100000     Structural::Model::Field#cast
                 100000.000      0.000      0.000 100000.000    100000/100000     <Module::Structural::Model::TypeCasts>#cast
                      0.000      0.000      0.000      0.000    100000/400000     Hash#fetch
--------------------------------------------------------------------------------
                 100000.000      0.000      0.000 100000.000    100000/100000     Structural::Model::Field#value_of
   8.62%   0.00% 100000.000      0.000      0.000 100000.000           100000     Structural::Model::Field#key
                 100000.000 100000.000      0.000      0.000    100000/300000     Symbol#to_s
                      0.000      0.000      0.000      0.000    100000/400000     Hash#fetch
--------------------------------------------------------------------------------
                  10000.000  10000.000      0.000      0.000      10000/10000     Structural::Hashifier#hashify
   0.86%   0.86%  10000.000  10000.000      0.000      0.000            10000     <Class::Hash>#[]
--------------------------------------------------------------------------------
                     10.000     10.000      0.000      0.000    100000/100000     Structural::Model#memoize
   0.00%   0.00%     10.000     10.000      0.000      0.000           100000     Kernel#instance_variable_set
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     <Module::Structural::Model::TypeCasts>#cast
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Structural::Model::TypeCasts::Identity#cast
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Class#new
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Structural::Model::TypeCasts::Cast#initialize
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     <Module::Structural::Model::TypeCasts>#cast
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     <Module::Structural::Model::TypeCasts>#casts
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/400000     Structural::Model::Field#key
                      0.000      0.000      0.000      0.000    100000/400000     Structural::Model::Field#value_of
                      0.000      0.000      0.000      0.000    100000/400000     Structural::Model::Field#cast
                      0.000      0.000      0.000      0.000    100000/400000     <Module::Structural::Model::TypeCasts>#cast
   0.00%   0.00%      0.000      0.000      0.000      0.000           400000     Hash#fetch
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model#memoize
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Kernel#instance_variable_get
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    300000/300000     Structural::Hashifier#as_data
   0.00%   0.00%      0.000      0.000      0.000      0.000           300000     Module#===
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Hash#each
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Structural::Hashifier#as_data
                      0.000      0.000      0.000      0.000    300000/300000     Module#===
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Hash#each
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     String#to_s
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000      10000/10000     Class#new
   0.00%   0.00%      0.000      0.000      0.000      0.000            10000     Structural::Hashifier#initialize

* indicates recursively called methods
```

By looking at this output, I was able to identify which parts on the code allocate excessive amount of objects and apply changes that reduced the number of allocations by `98%`. Here is the output of the profiles after all changes have been applied:

```
Measure Mode: allocations
Thread ID: 6699360
Fiber ID: 14254920
Total Time: 20031.0
Sort by: total_time

  %total   %self      total       self       wait      child            calls     name
--------------------------------------------------------------------------------
 100.00%   0.00%  20031.000      1.000      0.000  20030.000                1     Global#[No method]
                  20000.000  10000.000      0.000  10000.000      10000/10000     Class#new
                     30.000      0.000      0.000     30.000    100000/100000     Structural::Model#memoize
--------------------------------------------------------------------------------
                  20000.000  10000.000      0.000  10000.000      10000/10000     Global#[No method]
  99.85%  49.92%  20000.000  10000.000      0.000  10000.000            10000     Class#new
                  10000.000      0.000      0.000  10000.000      10000/10000     Structural::Model#initialize
--------------------------------------------------------------------------------
                  10000.000  10000.000      0.000      0.000      10000/10000     Structural::Model#initialize
  49.92%  49.92%  10000.000  10000.000      0.000      0.000            10000     <Class::Structural::Hashifier>#hashify
                      0.000      0.000      0.000      0.000      10000/10000     Kernel#tap
--------------------------------------------------------------------------------
                  10000.000      0.000      0.000  10000.000      10000/10000     Class#new
  49.92%   0.00%  10000.000      0.000      0.000  10000.000            10000     Structural::Model#initialize
                  10000.000  10000.000      0.000      0.000      10000/10000     <Class::Structural::Hashifier>#hashify
--------------------------------------------------------------------------------
                     30.000      0.000      0.000     30.000    100000/100000     Global#[No method]
   0.15%   0.00%     30.000      0.000      0.000     30.000           100000     Structural::Model#memoize
                     20.000      0.000      0.000     20.000    100000/100000     Proc#call
                     10.000     10.000      0.000      0.000    100000/100000     Kernel#instance_variable_set
                      0.000      0.000      0.000      0.000    100000/100000     Kernel#instance_variable_get
--------------------------------------------------------------------------------
                     20.000     20.000      0.000      0.000            10/10     Structural::Model::Field#default_value
   0.10%   0.10%     20.000     20.000      0.000      0.000               10     Kernel#proc
--------------------------------------------------------------------------------
                     20.000      0.000      0.000     20.000    100000/100000     Structural::Model::Field#value_of
   0.10%   0.00%     20.000      0.000      0.000     20.000           100000     Structural::Model::Field#default_value
                     20.000     20.000      0.000      0.000            10/10     Kernel#proc
--------------------------------------------------------------------------------
                     20.000      0.000      0.000     20.000    100000/100000     Proc#call
   0.10%   0.00%     20.000      0.000      0.000     20.000           100000     Structural::Model::Field#value_of
                     20.000      0.000      0.000     20.000    100000/100000     Structural::Model::Field#default_value
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model::Field#cast
                      0.000      0.000      0.000      0.000    100000/300000     Hash#fetch
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model::Field#key
--------------------------------------------------------------------------------
                     20.000      0.000      0.000     20.000    100000/100000     Structural::Model#memoize
   0.10%   0.00%     20.000      0.000      0.000     20.000           100000     Proc#call
                     20.000      0.000      0.000     20.000    100000/100000     Structural::Model::Field#value_of
--------------------------------------------------------------------------------
                     10.000     10.000      0.000      0.000    100000/100000     Structural::Model#memoize
   0.05%   0.05%     10.000     10.000      0.000      0.000           100000     Kernel#instance_variable_set
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     <Module::Structural::Model::TypeCasts>#cast
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Kernel#is_a?
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model::Field#cast
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     <Module::Structural::Model::TypeCasts>#cast
                      0.000      0.000      0.000      0.000    100000/100000     Kernel#is_a?
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model::Field#cast
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Kernel#class
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model::Field#value_of
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Structural::Model::Field#cast
                      0.000      0.000      0.000      0.000    100000/100000     <Module::Structural::Model::TypeCasts>#cast
                      0.000      0.000      0.000      0.000    100000/300000     Hash#fetch
                      0.000      0.000      0.000      0.000    100000/100000     Kernel#class
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model::Field#key
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Symbol#to_sym
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/300000     Structural::Model::Field#key
                      0.000      0.000      0.000      0.000    100000/300000     Structural::Model::Field#value_of
                      0.000      0.000      0.000      0.000    100000/300000     Structural::Model::Field#cast
   0.00%   0.00%      0.000      0.000      0.000      0.000           300000     Hash#fetch
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model::Field#value_of
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Structural::Model::Field#key
                      0.000      0.000      0.000      0.000    100000/100000     Symbol#to_sym
                      0.000      0.000      0.000      0.000    100000/300000     Hash#fetch
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Structural::Model#memoize
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     Kernel#instance_variable_get
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    300000/300000     <Class::Structural::Hashifier>#as_data
   0.00%   0.00%      0.000      0.000      0.000      0.000           300000     Module#===
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Hash#each
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     <Class::Structural::Hashifier>#as_data
                      0.000      0.000      0.000      0.000    300000/300000     Module#===
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000    100000/100000     Hash#each
   0.00%   0.00%      0.000      0.000      0.000      0.000           100000     String#to_sym
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000      10000/10000     Kernel#tap
   0.00%   0.00%      0.000      0.000      0.000      0.000            10000     Hash#each
                      0.000      0.000      0.000      0.000    100000/100000     <Class::Structural::Hashifier>#as_data
                      0.000      0.000      0.000      0.000    100000/100000     String#to_sym
--------------------------------------------------------------------------------
                      0.000      0.000      0.000      0.000      10000/10000     <Class::Structural::Hashifier>#hashify
   0.00%   0.00%      0.000      0.000      0.000      0.000            10000     Kernel#tap
                      0.000      0.000      0.000      0.000      10000/10000     Hash#each

* indicates recursively called methods
```

Additionally, this change reduced the time spend on executing the code above from by 40%. 

For more details regarding each change I've made, please have a look at individual commits.

### CODE REVIEW

File: lib/structural/model/type_casts.rb [hunk 1]
```diff
@@ -17,92 +20,71 @@ def self.casts
...
 
-      class Time < Cast
+      class Time
         def self.type
           ::Time
         end
 
-        def conversion
+        def self.cast(value)
+          puts value.inspect
```
Comment (seenmyfate) [inline lib/structural/model/type_casts.rb]: Should this still be here?

Comment (kajetanowicz) [inline lib/structural/model/type_casts.rb]: Upps

**Comment by carljohnstone** [comment]
:shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit:

**Comment by carljohnstone** [comment]
Are the cast changes going to change back-compat?

**Comment by kajetanowicz** [comment]
Yes, the external behaviour hasn't changed - I've just made some implementation changes.

**Pull request merged**